### ISSUE TITLE
Automatically cancel running queries

### ISSUE THREAD

**Issue by steven-sheehy** [pgjdbc/r2dbc-postgresql#251]
## Feature Request

Following up on #246, it would be nice to auto cancel long running queries when the subscription is cancelled. It looks like the PR only added an API to allow queries to be cancelled but didn't hook it into the flow to automatically cancel queries. This is a formal request for the latter scenario.

The following test case fails on master with #246 but should succeed:

```java
    @Test
    void autoCancelRequest() {
        this.connection.createStatement("SELECT pg_sleep(1000)")
                .execute()
                .as(StepVerifier::create)
                .thenCancel()
                .verify(Duration.ofMillis(500));

        this.connection.createStatement("SELECT * from pg_stat_activity where state = 'active' and query not like '%pg_stat_activity%'")
                .execute()
                .flatMap(PostgresqlResult::getRowsUpdated)
                .delaySubscription(Duration.ofSeconds(1))
                .as(StepVerifier::create)
                .expectNext(0)
                .verifyComplete();
    }
```

#### Is your feature request related to a problem? Please describe

Not cancelling a long running query can increase database utilization and exhaust the connection pool.

#### Describe the solution you'd like

When a user cancels the query (e.g. via `Flux.take(5)`, `Flux.limitRequest(5)`, etc.), R2DBC should explicitly request PostgreSQL to cancel the query's process ID instead of keeping it open and discarding subsequent messages as it does now via the `FluxDiscardOnCancel` operator.

#### Describe alternatives you've considered

Right now we always use limit with keyset pagination and have to poll to avoid this, which is sub-optimal.

**Comment by mp911de** [comment]
I'm not sure about the implications on small result sets within a highly concurrent environment. The driver makes use of pipelining which blurs the line of which statement gets actually canceled. Cancellation happens on a regular basis. (e.g. `connection.createStatement("INSERT INTO RETURNING *").flatMap(extractKey).first()`). Statements of this kind are completed already and cancellation would affect the statement that was sent after this one.

For larger, un-cursored results or long-running tasks, cancellation makes indeed sense to prevent unnecessary resource allocation. We introduced recently support for fetch sizes via #245 that allow cancelation by closing the portal when reaching the end of the current cursor chunk.

Canceling server-side requests makes only sense in the context of timeouts and with the intent to avoid a significant load. We cannot distinguish the latter `cancel()` signal from `next()` or `take(5)` therefore we cannot react appropriately.

The PG JDBC driver cancels a query if it's not yet completed but in that case, all calls to `ResultSet` are synchronous and await completion. `cancel()` unsubscribes from the upstream `Publisher` without awaiting completion.

**Comment by steven-sheehy** [comment]
It may be that using fetch size may be sufficient for my use. But it's currently impractical for use due to missing:

- Fetch size support in `DatabaseClient`
- Fetch size support in `@Query` methods
- Default fetch size in `ConnectionFactory`

I still think there's a need to cancel un-cursored queries. With PG JDBC you can cancel the statement and if the ResultSet is still awaiting completion you can always close the connection to force it to stop.

I don't understand enough about how this driver works to argue your points. Maybe it's naive, but if you have the pid of a query being ran on PostgreSQL and you receive a cancel signal, then it seems pretty straightforward that you'd want to tell the server to `pg_terminate_backend(pid)`. This cancellation can be a best effort. Whether the server returns a success that it cancelled it or if the server already terminated that pid and returns an error, either response can both be ignored and all future data can be discarded (as it does now).

**Comment by mp911de** [comment]
We could enable this behavior and run a couple tests with Spring Data to see how this goes. My statement is based on the typical effects of concurrency and probably I’m a bit overcautious.

**Comment by mp911de** [comment]
We introduced with #121 and #246 a possibility to cancel queries. Cursored queries are canceled (not continued) if the result consumption is being stopped and we introduced statement timeouts. I propose to close this ticket for the time being until we identify a good case where it makes sense to cancel queries in the backend based on the `cancel` signal because introducing such functionality requires refinements to a lot of areas.

**Comment by ndcotta-bc** [comment]
@mp911de I am hoping to reopen this debate and encourage you to consider to cancel queries when receiving the `cancel()` signal. My use motivation is twofold:
- **Expectations**. The reactivestreams javadoc for `Subscription` states that _"it is used to both signal desire for data and cancel demand (and allow resource cleanup)."_ When a user calls `cancel()` on a Subscription provided by R2DBC, I think it is not unreasonable for them to expect the query to get cancelled. Other libraries rely on this API contract too: Kotlin's coroutines (which can be made from a `Publisher<T>`) can be cancelled. I would expect a cancelled coroutine to propagate its cancellation, as is the case with other libraries.
- **API surface**. This is more specific to my use-case. I am using [r2dbc-pool](https://github.com/r2dbc/r2dbc-pool) (great lib!) to pool my R2DBC connections. It does not expose the underlying `PostgresqlConnection`, it only bothers implementing R2DBC's `Connection`, which is not affected by [#246](https://github.com/pgjdbc/r2dbc-postgresql/pull/246) . Therefore I cannot really call `cancelRunningQuery()` _and_ use r2dbc-pool. This is an example of how [#246](https://github.com/pgjdbc/r2dbc-postgresql/pull/246) does not propagate to consumer libs that are used, in practice, to perform queries (jOOQ is another good example of this). But they do use the `cancel()` API!

**Comment by ndcotta-bc** [comment]
@mp911de It's been a while since my last comment. Happy to open a new issue if that helps? If you don't think this is the right approach I would love to hear some more of your reasoning (you're the expert here after all).

This kind of change that modifies default behaviour for most paths is usually best done as soon as possible in the lifetime of the framework, in order to avoid breaking downstream if someone relies on the fact that cancellations are _not_ propagated to the DB.

**Comment by mp911de** [comment]
The only way how cancellation could work is by introducing a non-pipelining mode

> The driver makes use of pipelining which blurs the line of which statement gets actually canceled

Pipelining sends all requests to the server and without us consuming all response from the response socket, we cannot know, at what processing state the server has arrived. The flip side is, that we do not want to read all data from the socket into the JVM as that contradicts with the reactive (backpressure) way of handling I/O.

**Comment by ndcotta-bc** [comment]
I am with @steven-sheehy in that I do not fully understand how pipelining affects the driver and I defer to you on this,
but is the `pid` not (likely to be) available after the query starts?
> Maybe it's naive, but if you have the pid of a query being ran on PostgreSQL and you receive a cancel signal, then it seems pretty straightforward that you'd want to tell the server to pg_terminate_backend(pid). This cancellation can be a best effort.

If the naive part is that you do _not_ in fact have the `pid`, that is a different problem. Surely at some point in the execution it becomes available, since in #246 [it is used here](https://github.com/pgjdbc/r2dbc-postgresql/pull/246/files#diff-74dda26cb43d03521a2d93bd81e30da0d53adbd0b9f8747c6318da11af60eb9aR454).

I also agree with @steven-sheehy that this could be best effort:
- if the `pid` is _not_ yet available, then either
    - schedule a cancellation for when it becomes available OR
    - give up on trying to cancel this? I am open to suggestions
- if it is, then an unsuccessful response (or a closed connection) can be ignored
